system/jlib/jutil.cpp

```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
+    }
+    bool requestSecretByCategory(const char *category, CVaultKind &kind, StringBuffer &content, const char *secret, const char *version) override
+    {
+        if (isEmptyString(category))
+            return false;
+        std::map<std::string, std::unique_ptr<CVaultSet>>::iterator it = categories.find(category);
+        if (it == categories.end())
+            return false;
+        return it->second->requestSecret(kind, content, secret, version);
+    }
+};
+
+static Owned<CVaultManager> vaultManager;
+IVaultManager *ensureVaultManager()
+{
+    CriticalBlock block(secretCS);
+    if (!vaultManager)
+        vaultManager.setown(new CVaultManager());
+    return vaultManager;
+}
+
+CriticalSection secretCacheCS;
+static Owned<IPropertyTree> secretCache = createPTree();
+
+static IPropertyTree *getCachedLocalSecret(const char *name)
+{
+    if (isEmptyString(name))
+        return nullptr;
+    unsigned timeoutThreshold = msTick() - getSecretTimeout();
+    Owned<IPropertyTree> secret;
+    {
+        CriticalBlock block(secretCacheCS);
+        secret.setown(secretCache->getPropTree(name));
+        if (secret)
+        {
+            unsigned created = (unsigned) secret->getPropInt("@created");
+            if (created && (created < timeoutThreshold))
+            {
+                secretCache->removeProp(name);
+                puts("\nremoved from local cache\n");
+                return nullptr;
+            }
+            puts("\nretrieved from local cache\n");
+            return secret.getClear();
+        }
+    }
+    return nullptr;
+}
+
+static void addCachedLocalSecret(const char *name, IPropertyTree *secret)
+{
+    if (!secret || isEmptyString(name))
+        return;
+    secret->setPropInt("@created", (int)msTick());
+    {
+        CriticalBlock block(secretCacheCS);
+        secretCache->setPropTree(name, LINK(secret));
+        puts("\nadded to local cache\n");
+    }
+}
+
+static const char *ensureSecretDirectory()
+{
+    CriticalBlock block(secretCS);
+    if (secretDirectory.isEmpty())
+        setSecretMount(nullptr);
+    return secretDirectory;
+}
+
+static IPropertyTree *loadLocalSecret(const char * name)
+{
     StringBuffer path;
-    addPathSepChar(path.append(secretDirectory)).append(name).append(PATHSEPCHAR).append(key);
-    Owned<IFile> file = createIFile(path);
-    result.loadFile(file);
-    return result;
+    addPathSepChar(path.append(ensureSecretDirectory())).append(name).append(PATHSEPCHAR);
+    Owned<IDirectoryIterator> dir = createDirectoryIterator(path);
+    if (!dir || !dir->first())
+        return nullptr;
+    Owned<IPropertyTree> tree = createPTree(name);
+    tree->setPropInt("@created", (int) msTick());
+    ForEach(*dir)
+    {
+        StringBuffer name;
+        dir->getName(name);
+        if (!validateXMLTag(name))
+            continue;
+        StringBuffer value;
+        value.loadFile(&dir->query());
+        tree->setProp(name, value.str());
```

Comment:
What implications does this have for binary secrets?  Is it ok to assume they will never be used?

Reply:
I'll do a little research.